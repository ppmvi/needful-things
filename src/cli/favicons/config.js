import fs from 'fs';
import path from 'path';

const filePath = path.resolve(process.cwd(), './.ppmvi.js');
let userConfig;

if (fs.existsSync(filePath)) {
  const file = require(filePath).default;
  userConfig = file.favicons || {};
}

if (userConfig && userConfig.master_picture && userConfig.master_picture.type === 'inline') {
  const image = path.resolve(process.cwd(), userConfig.master_picture.content);
  if (fs.existsSync(image)) {
    userConfig.master_picture.content = fs.readFileSync(image).toString('base64');
  }
}

const defaultConfig = {
  'api_key': '',
  'master_picture': {
    'type': 'url',
    'url': ''
  },
  'files_location': {
    'type': 'path',
    'path': '/static/icons'
  },
  'favicon_design': {
    'desktop_browser': {},
    'ios': {
      'picture_aspect': 'background_and_margin',
      'margin': '8',
      'background_color': '#FFFFFF',
      'assets': {
        'ios6_and_prior_icons': false,
        'ios7_and_later_icons': false,
        'precomposed_icons': false,
        'declare_only_default_icon': true
      }
    },
    'windows': {
      'picture_aspect': 'white_silhouette',
      'background_color': '#FFFFFF',
      'assets': {
        'windows_80_ie_10_tile': true,
        'windows_10_ie_11_edge_tiles': {
          'small': false,
          'medium': true,
          'big': false,
          'rectangle': false
        }
      }
    },
    'android_chrome': {
      'picture_aspect': 'background_and_margin',
      'margin': '12',
      'background_color': '#FFFFFF',
      'theme_color': '#FFFFFF',
      'manifest': {
        'name': 'My sample app',
        'display': 'standalone',
        'orientation': 'portrait',
        'start_url': '/'
      },
      'assets': {
        'legacy_icon': false,
        'low_resolution_icons': false
      }
    },
    'safari_pinned_tab': {
      'picture_aspect': 'black_and_white',
      'threshold': 60,
      'theme_color': '#FFFFFF'
    }
  },
  'settings': {
    'compression': '5',
    'scaling_algorithm': 'Mitchell',
    'error_on_image_too_small': true,
    'readme_file': false,
    'html_code_file': false,
    'use_path_as_is': true
  }
};

export default { ...defaultConfig, ...userConfig };
