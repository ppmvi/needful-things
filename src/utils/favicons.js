export function applyFaviconLinks(options = {}) {
  const {
    url = '/static/icons',
    manifest = {},
    color = '#FFFFFF'
  } = options;

  const {
    url: manifestUrl = '/static',
    name: manifestName = 'manifest.json'
  } = manifest;

  return [
    { rel: 'apple-touch-icon', sizes: '180x180', href: `${url}/apple-touch-icon.png` },
    { rel: 'icon', type: 'image/png', href: `${url}/favicon-32x32.png`, sizes: '32x32' },
    { rel: 'icon', type: 'image/png', href: `${url}/favicon-16x16.png`, sizes: '16x16' },
    { rel: 'manifest', href: `${manifestUrl}/${manifestName}` },
    { rel: 'mask-icon', href: `${url}/safari-pinned-tab.svg`, color: color },
    { rel: 'shortcut icon', href: `${url}/favicon.ico` }
  ];
}

export function applyFaviconMetaTags(options = {}) {
  const {
    url = '/static/icons',
    color = '#FFFFFF',
    statusBarStyle = 'default'
  } = options;

  return [
    { name: 'msapplication-TileColor', content: color },
    { name: 'msapplication-config', content: `${url}/browserconfig.xml` },
    { name: 'theme-color', content: color },
    { name: 'apple-mobile-web-app-status-bar-style', content: statusBarStyle },
    { name: 'apple-mobile-web-app-capable', content: 'yes' }
  ];
}
