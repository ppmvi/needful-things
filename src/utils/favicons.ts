import { LinkPropertyBase, LinkPropertyHref, MetaPropertyName } from 'vue-meta';

interface FaviconLinksOptions {
  path?: string;
  manifestPath?: string;
  color?: string;
}

interface FaviconMetaTagsOptions {
  path?: string;
  color?: string;
  statusBarStyle?: 'default' | 'black' | 'black-translucent';
}

export function applyFaviconLinks(
  options: FaviconLinksOptions = {}
): (LinkPropertyBase | LinkPropertyHref)[] {
  const {
    path = '/static/icons',
    manifestPath = '/static/manifest.json',
    color = '#FFFFFF',
  } = options;

  return [
    {
      rel: 'apple-touch-icon',
      sizes: '180x180',
      href: `${path}/apple-touch-icon.png`,
    },
    {
      rel: 'icon',
      type: 'image/png',
      href: `${path}/favicon-32x32.png`,
      sizes: '32x32',
    },
    {
      rel: 'icon',
      type: 'image/png',
      href: `${path}/favicon-16x16.png`,
      sizes: '16x16',
    },
    { rel: 'manifest', href: `${manifestPath}` },
    { rel: 'mask-icon', href: `${path}/safari-pinned-tab.svg`, color: color },
    { rel: 'shortcut icon', href: `${path}/favicon.ico` },
  ];
}

export function applyFaviconMetaTags(
  options: FaviconMetaTagsOptions = {}
): MetaPropertyName[] {
  const {
    path = '/static/icons',
    color = '#FFFFFF',
    statusBarStyle = 'default',
  } = options;

  return [
    { name: 'msapplication-TileColor', content: color },
    { name: 'msapplication-config', content: `${path}/browserconfig.xml` },
    { name: 'theme-color', content: color },
    { name: 'apple-mobile-web-app-status-bar-style', content: statusBarStyle },
    { name: 'apple-mobile-web-app-capable', content: 'yes' },
  ];
}
