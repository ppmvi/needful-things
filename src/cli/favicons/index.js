import rfg from 'rfg-api';
import pify from 'pify';
import config from './config';
import ora from 'ora';

export default async function generate(path = './static/icons') {
  const watcherSpinner = ora('Generating Favicons').start();
  const generator = rfg.init();
  
  try {
    await pify(generator.generateFavicon)(config, path);
    watcherSpinner.succeed();
  } catch (err) {
    watcherSpinner.fail(err.message);
  }
}
