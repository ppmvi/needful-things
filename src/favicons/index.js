import rfg from 'rfg-api';
import pify from 'pify';
import config from './config';
import ora from 'ora';

export async function generate() {
  const watcherSpinner = ora('Generating Favicons').start();
  const generator = rfg.init();
  
  try {
    await pify(generator.generateFavicon)(config, './icons');
    watcherSpinner.succeed();
  } catch (err) {
    watcherSpinner.fail(err.message);
  }
}
