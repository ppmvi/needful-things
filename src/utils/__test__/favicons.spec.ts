import * as favicons from '../favicons';

describe('favicons.ts', () => {
  test('applyFaviconLinks should match snapshot with default values', () => {
    const links = favicons.applyFaviconLinks();
    expect(links).toMatchSnapshot();
  });

  test('applyFaviconLinks should match snapshot with given values', () => {
    const links = favicons.applyFaviconLinks({
      path: '/test',
      manifestPath: '/test/test.json',
      color: '#000',
    });
    expect(links).toMatchSnapshot();
  });

  test('applyFaviconMetaTags should match snapshot with default values', () => {
    const links = favicons.applyFaviconMetaTags();
    expect(links).toMatchSnapshot();
  });

  test('applyFaviconMetaTags should match snapshot with given values', () => {
    const links = favicons.applyFaviconMetaTags({
      color: '#000',
      path: '/test',
      statusBarStyle: 'black-translucent',
    });
    expect(links).toMatchSnapshot();
  });
});
