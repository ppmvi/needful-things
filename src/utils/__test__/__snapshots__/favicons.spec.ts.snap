// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`favicons.ts applyFaviconLinks should match snapshot with default values 1`] = `
Array [
  Object {
    "href": "/static/icons/apple-touch-icon.png",
    "rel": "apple-touch-icon",
    "sizes": "180x180",
  },
  Object {
    "href": "/static/icons/favicon-32x32.png",
    "rel": "icon",
    "sizes": "32x32",
    "type": "image/png",
  },
  Object {
    "href": "/static/icons/favicon-16x16.png",
    "rel": "icon",
    "sizes": "16x16",
    "type": "image/png",
  },
  Object {
    "href": "/static/manifest.json",
    "rel": "manifest",
  },
  Object {
    "color": "#FFFFFF",
    "href": "/static/icons/safari-pinned-tab.svg",
    "rel": "mask-icon",
  },
  Object {
    "href": "/static/icons/favicon.ico",
    "rel": "shortcut icon",
  },
]
`;

exports[`favicons.ts applyFaviconLinks should match snapshot with given values 1`] = `
Array [
  Object {
    "href": "/test/apple-touch-icon.png",
    "rel": "apple-touch-icon",
    "sizes": "180x180",
  },
  Object {
    "href": "/test/favicon-32x32.png",
    "rel": "icon",
    "sizes": "32x32",
    "type": "image/png",
  },
  Object {
    "href": "/test/favicon-16x16.png",
    "rel": "icon",
    "sizes": "16x16",
    "type": "image/png",
  },
  Object {
    "href": "/test/test.json",
    "rel": "manifest",
  },
  Object {
    "color": "#000",
    "href": "/test/safari-pinned-tab.svg",
    "rel": "mask-icon",
  },
  Object {
    "href": "/test/favicon.ico",
    "rel": "shortcut icon",
  },
]
`;

exports[`favicons.ts applyFaviconMetaTags should match snapshot with default values 1`] = `
Array [
  Object {
    "content": "#FFFFFF",
    "name": "msapplication-TileColor",
  },
  Object {
    "content": "/static/icons/browserconfig.xml",
    "name": "msapplication-config",
  },
  Object {
    "content": "#FFFFFF",
    "name": "theme-color",
  },
  Object {
    "content": "default",
    "name": "apple-mobile-web-app-status-bar-style",
  },
  Object {
    "content": "yes",
    "name": "apple-mobile-web-app-capable",
  },
]
`;

exports[`favicons.ts applyFaviconMetaTags should match snapshot with given values 1`] = `
Array [
  Object {
    "content": "#000",
    "name": "msapplication-TileColor",
  },
  Object {
    "content": "/test/browserconfig.xml",
    "name": "msapplication-config",
  },
  Object {
    "content": "#000",
    "name": "theme-color",
  },
  Object {
    "content": "black-translucent",
    "name": "apple-mobile-web-app-status-bar-style",
  },
  Object {
    "content": "yes",
    "name": "apple-mobile-web-app-capable",
  },
]
`;
